docs/CodingStandards.md

```diff
@@ -0,0 +1,162 @@
+
+  Further exceptions outlined in the LLVM coding standard are where a class has a method that
+  mimics an STL class action, such as `push_back`. Such a method can keep the STL-like name.
+
+* A *constant* (technically a static variable or class/struct member variable, but `final` so it is used
+  like a constant) should be in CamelCase.
+
+* An *enumerator* (e.g. `enum { Foo, Bar }`) acts like a constant, so should start with an
+  upper-case letter. In addition, if the enum is not an enum class, and is not hidden inside a
+  class that suitably qualifies it, then each enumerator should have
+  a prefix corresponding to (not necessarily exactly the same as) the enum declaration name.
+
+Unlike in LLVM, an initialism (e.g. "NGG") should be treated as a word in identifier-naming
+rules. Thus, if it appears at the start of a variable or function name, it should be entirely
+lower-case (`nggControl`); if it appears at the start of a type or constant name, or anywhere
+other than the start in any name, it should have an initial upper-case letter (`NggControl`).
+
+For API compatibility, llpc.h has not been modified from PAL-style naming, which means it still
+contains struct members with a p prefix for pointer, and CamelCase for methods. Any method
+within LLPC that overrides a method in a class declared in llpc.h, such as an ICompiler method,
+also keeps CamelCase rather than camelBack.
+
+### Braces
+
+Brace placement is covered by clang-format below.
+
+C++ allows the body of an if, else, while, do, for to be a single simple statement,
+without any braces. We follow the LLVM convention of only using an unbraced body if
+it is a single line, including comments. Otherwise the body should be braced.
+
+### Redundant comparisons
+
+Comparison of a bool with `false` or a pointer with `nullptr` is redundant and should
+be avoided. Use `!` instead where appropriate.
+
+However a comparison of an integer with 0 is useful. Consider keeping that.
+
+### Redundant parentheses
+
+Certain cases of theoretically redundant parentheses around a subexpression are needed
+because compilers warn about them otherwise, for the good reason that C/C++ operator
+precedence makes it difficult to understand and easy to get wrong (in particular
+`&` `|` `^` vs comparison operators, and `&&` vs `||`). However redundant parentheses
+are not needed in cases that compilers do not warn about, in particular a comparison
+inside `&&` or `||`, or a comparison used as the condition of `?:` or in an assignment
+or `return`. Avoid redundant parentheses in these cases, unless really needed for
+clarity.
+
+### Comments
+
+Despite the look of much LLVM code, the LLVM coding standard does actually recommend
+placing a comment at the top of a function/method. :-)
+
+In LLPC, that is mandatory. Parameters should be commented in the main comment at
+the top of the function like this:
+
+```
+// @param hatstand : [in/out] The Hatstand object to modify
+```
```

Comment:
IMO there's little value in comments that basically repeat what the code says in a trivial way. Consider the example you gave; in code, it'll look something like this:

```
// @param hatstand : [in/out] The Hatstand object to modify.
// @param newName:  The new name.
void foo(Hatstand &hatstand, const std::string &newName) {
```
I think it's much more reasonable to use good argument names in the first place, and if the use of the parameter is not obvious, then give a higher-level explanation that connects this parameter with the function contract/logic.

Otherwise you end up writing and *having to maintain* the same thing twice for little benefit, which becomes basically this very quickly:
![image](https://user-images.githubusercontent.com/4612584/75561832-a0fd7680-5a15-11ea-80cf-b4ea93dcacab.png)

Reply:
Initially, this was discussed if we should let some parameter comments just be vacant since the names had already shown how to use them. Later, we find this will lead developers less likely to do commenting because everyone could believe "there is no need of making comments for this or these parameters. They are so obvious". Hence, we require each parameter should have its own comments even if the words or sentences are trivial. This is also the reviewers' responsibilities to help code writer to better describe what they want to tell code readers especially for complex or tricky use of a special parameter.

Reply:
I agree with Rex here. While it's true that often, the parameter descriptions are ideally redundant, having this rule forces both the original writer of the code and the reviewer to think about whether there isn't something more interesting to comment about. Often, there is.